### PR TITLE
TensorArray demo notebook

### PR DESCRIPTION
**Pull request by BryanCutler**
This notebook uses parts of the CoNLL-2003 tutorial to showcase the use of TensorArray to represent BERT embeddings for model training and prediction results.

### CODE REVIEW

File: tutorials/corpus/util.py [hunk 1]
```diff
@@ -73,6 +74,8 @@ def predict_on_df(df: pd.DataFrame, id_to_class: Dict[int, str], predictor):
     iobs, types = tp.decode_class_labels(result_df["predicted_class"].values)
     result_df["predicted_iob"] = iobs
     result_df["predicted_type"] = types
+    prob_values = predictor.predict_proba(x_values)
+    result_df["predicted_class_pr"] = tp.TensorArray(prob_values)
```
Comment (BryanCutler) [inline tutorials/corpus/util.py:78]: I couldn't use this function from my notebook, I just made a copy of it. Is it something we want to keep updated with this?

Comment (BryanCutler) [reply tutorials/corpus/util.py:78]: TODO: The notebook version is slightly different now

**Comment by BryanCutler** [comment]
cc @frreiss to please take a look, also @kmh4321 and @MLnick if interested.

**Comment by BryanCutler** [comment]
@frreiss I had a problem in the notebook trying to align a span column with the token_id of the BERT DataFrame. The span column gets dropped with the DataFrame is saved, but I wanted to show something at the end, but I seem to be missing something to add back in the span column.

**Comment by MLnick** [comment]
@BryanCutler where is the notebook itself? Doesn't seem part of the PR currently.

**Comment by BryanCutler** [comment]
Oops, sorry @MLnick , it would help if I added the file. Doing it now..

File: tutorials/corpus/util.py (hunk 1 above)
Comment (frreiss) [reply tutorials/corpus/util.py:78]: Having multiple copies of this function is ok. If this code was core functionality, I would have put this function into the library proper instead of leaving it in a directory of notebooks.

**Review by frreiss — CHANGES_REQUESTED** [review]
Still not seeing the notebook itself in these commits. Did you forget to add a file?

**Comment by frreiss** [comment]
Ok, now the commit with the notebook is showing up. Strange. Will review momentarily.

File: notebooks/Model_Training_with_TensorArray.ipynb [hunk 2]
```diff
@@ -0,0 +1,5826 @@
...
+       "[281109 rows x 17 columns]"
+      ]
+     },
+     "execution_count": 39,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "train_results_df = predict_on_df(train_df, int_to_label, base_model)\n",
```
Comment (frreiss) [inline notebooks/Model_Training_with_TensorArray.ipynb:3757]: I don't think it makes much sense to run inference on the training set. I recommend you skip ahead to running inference on the test set.

File: notebooks/Model_Training_with_TensorArray.ipynb [hunk 3]
```diff
@@ -0,0 +1,5826 @@
...
+      ]
+     },
+     "execution_count": 42,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Look at results on the test set\n",
+    "test_results_df = predict_on_df(corpus_df[corpus_df[\"fold\"] == \"test\"], \n",
```
Comment (frreiss) [inline notebooks/Model_Training_with_TensorArray.ipynb:4080]: These results contain label probabilities for each of the BERT tokens in the test fold of the corpus. To obtain predictions of the named entities in the test fold, you need to convert these probabilities to one "best" label (IOB tag + entity type) per token. The easiest way to get those "best" labels is to take the highest probability label at each token position. Then you 'll want to convert those token labels to (span, entity type pairs), the same way that you processed the corpus itself earlier in this notebook. Once you've made that conversion, you can compute precision and recall a la Analyze_Model_Outputs.ipynb.

**Review by frreiss — APPROVED** [review]
This is looking good. Could use some additional explanations at the beginning and some additional postprocessing/scoring code at the end (see comments throughout), but this is in good enough shape to merge as-is for now.

**Comment by BryanCutler** [comment]
Thanks for reviewing @frreiss , I'll work on an update

File: notebooks/Model_Training_with_TensorArray.ipynb [hunk 4]
```diff
@@ -0,0 +1,5826 @@
...
+    "\n",
+    "A `TensorArray` can be constructed with zero copy from a single `numpy.ndarray` or with a sequence of elements of similar shaprs. Conversion of a `TensorArray` to a `numpy.ndarray` can be done with zero copy by calling `TensorArray.to_numpy()` or using the provided numpy array interface, e.g. `numpy.asarray(TensorArray(...))`. The `TensorArray` is a Pandas extension type of type `TensorType` and can be wrapped in a `pandas.Series` or used as a column in a `pandas.DataFrame` and used in standard Pandas operations. A `NULL` or missing value in the `TensorArray` is represented as a N-dimensional `numpy.ndarray` where all items are `numpy.nan`.\n",
+    "\n",
+    "## TensorArray Example Usage\n",
+    "\n",
+    "To demostrate the TensorArray functionality from Text Extensions for Pandas, we will use the CoNLL-2003 dataset [TODO REF] to \n",
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply notebooks/Model_Training_with_TensorArray.ipynb:28]: Thanks @frreiss , I'll add that. Should I reference the other more complete CoNLL tutorial notebooks we have also?

File: notebooks/Model_Training_with_TensorArray.ipynb [hunk 5]
```diff
@@ -0,0 +1,5826 @@
+{
```
Comment (MLnick) [inline notebooks/Model_Training_with_TensorArray.ipynb:1]: <p>"Standard arithetic and comparison ..." -&gt; "Standard <strong>arithmetic</strong> and comparison ..."</p><p> </p><p>"... elements of similar shaprs." -&gt; "...  elements of similar shapes."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (MLnick) [inline notebooks/Model_Training_with_TensorArray.ipynb:1]: <p>In <code>spans_df = spans_df.reindex(columns=mns=["span", "token_span", "ent_type"])</code>  is it supposed to be <code>columns=mns=...</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

**Comment by MLnick** [comment]
Just a couple minor things picked up so far - still working through the technical details of the notebook. Though agreed with @frreiss that some more description of the problem & goal etc would be useful.

File: notebooks/Model_Training_with_TensorArray.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (frreiss) [reply notebooks/Model_Training_with_TensorArray.ipynb:28]: The stuff in the `tutorials` directory is currently a bit half-baked (also somewhat subject to anonymity constraints). I recommend including a reference to the other CoNLL-related notebook in this directory, though -- `Analyze_Model_Outputs.ipynb`

**Comment by BryanCutler** [comment]
Thanks @MLnick , I will reword things to better state the problem & goal.

**Comment by BryanCutler** [comment]
@frreiss and @MLnick , I updated a lot of the wording and added more descriptions and links. Hopefully, this better states the problem and goal of the notebook. Please take another look when you can, thanks!

File: text_extensions_for_pandas/array/test_tensor.py [hunk 6]
```diff
@@ -345,12 +345,35 @@ def test_sum(self):
...
     def test_factorize(self):
         x = np.array([[1, 2], [3, 4], [5, 6], [3, 4]])
         s = TensorArray(x)
         with self.assertRaises(NotImplementedError):
             indices, values = s.factorize()
 
+    def test_take(self):
+        # Test no missing gets same dtype
+
+        # Test missing with nan fill gets promoted to float
```
Comment (BryanCutler) [inline text_extensions_for_pandas/array/test_tensor.py:374]: TODO

File: text_extensions_for_pandas/io/bert.py [hunk 7]
```diff
@@ -216,7 +216,7 @@ def align_bert_tokens_to_corpus_tokens(
         [["corpus_token", "ent_type"]]
             .rename(columns={"corpus_token": "span"})
     )
-    cons_df["span"] = TokenSpanArray.align_to_tokens(
+    cons_df["token_span"] = TokenSpanArray.align_to_tokens(
```
Comment (BryanCutler) [inline text_extensions_for_pandas/io/bert.py:219]: @frreiss before this change it is replacing the existing "span" column, could you just verify that is what we want or should we do this to make a new "token_span" column?

Comment (BryanCutler) [reply text_extensions_for_pandas/io/bert.py:219]: this is also the cause of the current test failure

**Comment by BryanCutler** [comment]
Not sure why the ReviewNB link wasn't added, but here it is

https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/

File: text_extensions_for_pandas/io/bert.py (hunk 7 above)
Comment (frreiss) [reply text_extensions_for_pandas/io/bert.py:219]: Returning both a token-based span and the original makes sense. Would you mind updating the description of the returned DataFrame's schema in the docstring above to match? If I'm reading this code correctly, the schema should be `[span, token_span, ent_type]`, where `span` is the original span from the input and `token_span` is a token-based span aligned to the BERT tokens.

**Review by frreiss — COMMENTED** [review]
Code changes outside the notebook look good.

File: notebooks/Model_Training_with_BERT.ipynb [hunk 8]
```diff
@@ -0,0 +1,5371 @@
+{
```
Comment (frreiss) [inline notebooks/Model_Training_with_BERT.ipynb:1]: <p>This cell can be merged with the next cell. Also, you can move the initialization of <code>bert</code>  down to the later cell where that object is first used.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (frreiss) [inline notebooks/Model_Training_with_BERT.ipynb:1]: <p>I recommend moving the code that initializes the variable <code>bert</code>  to this cell, so the reader doesn't have to scroll up to find it.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (frreiss) [inline notebooks/Model_Training_with_BERT.ipynb:1]: <p>Can you move the text of this comment to the markdown cell above and correct the spelling of "convenience"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (frreiss) [inline notebooks/Model_Training_with_BERT.ipynb:1]: <p>Can you add a computation of precision and recall at this point in the notebook, then redo the precision/recall computation after correcting for the BERT model splitting tokens? There should be a measurable increase in accuracy as a result of that correction.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (frreiss) [inline notebooks/Model_Training_with_BERT.ipynb:1]: <p>Take a look at Analyze_Model_Outputs.ipynb for an example of drilling down to problematic documents based on document-level precision and recall numbers.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

File: text_extensions_for_pandas/array/tensor.py [hunk 9]
```diff
@@ -224,14 +224,24 @@ def take(
...
-            values = np.full((len(indices),) + self._tensor.shape[1:], fill_value)
-            for i, idx in enumerate(indices):
-                if idx >= 0:
-                    values[i] = self._tensor[idx]
-        else:
-            values = self._tensor.take(indices, axis=0)
+
+            # Check if there are missing indices to fill, if not can use numpy take below
+            has_missing = np.any(indices < 0)
+            if has_missing:
```
Comment (BryanCutler) [inline text_extensions_for_pandas/array/tensor.py:230]: fyi, need this check so that it doesn't unnecessarily promote ints to floats if there are no missing values. Noticed this when sorting a DataFrame with TensorArray, which is now in a cell of the notebook in https://github.com/CODAIT/text-extensions-for-pandas/pull/122

File: notebooks/Model_Training_with_BERT.ipynb (hunk 8 above)
Comment (BryanCutler) [reply notebooks/Model_Training_with_BERT.ipynb:1]: <p>Yeah, that's a good idea. It does give it a little bump ~0.85 -&gt; ~0.87 for F1. @frreiss do you happen to know why the model sometimes splits tokens that way - is it an expected artifact or is something wrong?</p>

File: text_extensions_for_pandas/io/bert.py (hunk 7 above)
Comment (BryanCutler) [reply text_extensions_for_pandas/io/bert.py:221]: On second thought, adding an additional column "token_span" makes the notebook a bit more confusing and later functions are set to work with a "span" column, so it would need to be changed too.

I revert my changes so that the "span" column is replaced by a TokenSpanArray, as before, and updated the docs. I think it's clearer this way, wdyt?

File: notebooks/Model_Training_with_BERT.ipynb (hunk 8 above)
Comment (BryanCutler) [reply notebooks/Model_Training_with_BERT.ipynb:1]: I'll try to cover this as a followup
